### PR TITLE
[MU4] Palette step 5 - load from workspace, drag and drop

### PR DESCRIPTION
**Pull request by igorkorsukov**

### CODE REVIEW

File: framework/workspace/internal/workspace.cpp [hunk 1]
```diff
@@ -118,7 +120,7 @@ Ret Workspace::read()
 
 Ret Workspace::readWorkspace(const QByteArray& data)
 {
-    QXmlStreamReader xml(data);
+    Ms::XmlReader xml(data);
```
[the remark this answers is not in the document]
Comment (igorkorsukov) [reply framework/workspace/internal/workspace.cpp:123]: Unfortunately, the current implementation of reading palettes uses `Ms::XmlReader`, it’s difficult to change it, so now it’s easier to use `Ms::XmlReader`

File: mu4/scenes/palette/internal/paletteconfiguration.cpp [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+#include "settings.h"
+
+using namespace mu::scene::palette;
+using namespace mu::framework;
+
+static const Settings::Key PALETTE_SCALE("palette", "application/paletteScale");
+static const Settings::Key PALETTE_USESINGLE("palette", "application/useSinglePalette");
+
+void PaletteConfiguration::init()
```
[the remark this answers is not in the document]
Comment (igorkorsukov) [reply mu4/scenes/palette/internal/paletteconfiguration.cpp:31]: It will be needed later :)

File: mu4/scenes/notation/view/notationviewinputcontroller.cpp [hunk 3]
```diff
@@ -163,6 +166,75 @@ void NotationViewInputController::hoverMoveEvent(QHoverEvent* ev)
     }
 }
 
+void NotationViewInputController::dragEnterEvent(QDragEnterEvent* ev)
+{
+    //LOGI() << "ev: " << ev;
+
+    const QMimeData* dta = ev->mimeData();
```
Comment (Eism) [inline mu4/scenes/notation/view/notationviewinputcontroller.cpp:173]: Do you need nullptr checks here?

File: framework/global/ret.h [hunk 4]
```diff
@@ -33,7 +33,7 @@ class Ret
         UnknownError    = 1,
 
         // not error, just codes
-        Cancel       = 3,  // abort by user
```
Comment (vpereverzev) [inline framework/global/ret.h:36]: I think it would be useful to remove similar art indentation so as not to do it every single time this enum extends

File: framework/global/runtime.h [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+//  You should have received a copy of the GNU General Public License
+//  along with this program; if not, write to the Free Software
+//  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
+//=============================================================================
+#ifndef MU_FRAMEWORK_RUNTIME_H
+#define MU_FRAMEWORK_RUNTIME_H
+
+namespace mu {
+namespace runtime {
+inline constexpr bool IsDebug()
```
Comment (vpereverzev) [inline framework/global/runtime.h:24]: You can apparently use it in "ptrutils"

File: mu4/appshell/dockwindow/dockwindow.cpp [hunk 6]
```diff
@@ -322,3 +325,13 @@ void DockWindow::setCurrentPageUri(QString uri)
 
     emit currentPageUriChanged(m_currentPageUri);
 }
+
+QMainWindow* DockWindow::qMainWindow()
```
Comment (vpereverzev) [inline mu4/appshell/dockwindow/dockwindow.cpp:329]: Why not just mainWindow()?

File: framework/global/runtime.h (hunk 5 above)
Comment (igorkorsukov) [reply framework/global/runtime.h:24]: This requires C++ 17, but now unfortunately C++ 11

File: framework/global/ptrutils.h [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+#include "runtime.h"
+#include "log.h"
+
+namespace mu {
+namespace ptr {
+template<typename T, typename E> T* checked_cast(E* source)
+{
+#ifndef NDEBUG
+    T* casted = dynamic_cast<T*>(source);
```
[the remark this answers is not in the document]
Comment (igorkorsukov) [reply framework/global/ptrutils.h:30]: >> Is there any reason to not use only static_cast?

Because

```
    struct A
    {
        virtual ~A() = default;
    };

    struct B : public A
    {
    };

    struct C : public A
    {
    };

    A* c = new C();

    B* b_c = static_cast<B*>(c);
    B* b_c_dyn = dynamic_cast<B*>(c);

    LOGI() << "b_c: " << b_c << ", b_c_dyn: " << b_c_dyn;
```

output: `b_c:  0x561600215de0 , b_c_dyn:  0x0`

>> In general, I thought RTTI is disabled within the editor
As far as I can see - no, at least in the debug RTTI is turned on.
Although I don't understand why turn it off, an attempt to increase performance ... hmm.

File: mu4/appshell/dockwindow/dockwindow.cpp (hunk 6 above)
Comment (igorkorsukov) [reply mu4/appshell/dockwindow/dockwindow.cpp:329]: At the moment, this is a forced, I hope a temporary method, we need exactly `QMainWindow`.
I do not think that in this case, it is the main window in essence (from an architectural point of view)

File: mu4/scenes/notation/view/notationviewinputcontroller.cpp (hunk 3 above)
Comment (igorkorsukov) [reply mu4/scenes/notation/view/notationviewinputcontroller.cpp:173]: I don't know, nothing wrote about this in the documentation. I added check.

File: mu4/domain/notation/internal/notationinteraction.h [hunk 8]
```diff
@@ -96,23 +103,37 @@ class NotationInteraction : public INotationInteraction
 
     Ms::Page* point2page(const QPointF& p) const;
     QList<Element*> hitElements(const QPointF& p_in, float w) const;
+    QList<Element*> elementsAt(const QPointF& p) const;
+    Element* elementAt(const QPointF& p) const;
```
Comment (vpereverzev) [inline mu4/domain/notation/internal/notationinteraction.h:107]: So, what is the difference between **hitElements** and **elementsAt** methods?

File: mu4/domain/notation/notationdomainmodule.cpp [hunk 9]
```diff
@@ -47,6 +46,8 @@ void NotationDomainModule::registerExports()
     std::shared_ptr<INotationReadersRegister> readers = std::make_shared<NotationReadersRegister>();
     readers->reg({ "mscz", "mscx" }, std::make_shared<MsczNotationReader>());
     framework::ioc()->registerExport<INotationReadersRegister>(moduleName(), readers);
+
+    Notation::init();
```
Comment (vpereverzev) [inline mu4/domain/notation/notationdomainmodule.cpp:50]: Why did you move that?

File: mu4/scenes/palette/internal/paletteconfiguration.h [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+
+namespace mu {
+namespace scene {
+namespace palette {
+class PaletteConfiguration : public IPaletteConfiguration
+{
+public:
+    PaletteConfiguration() = default;
+
+    void init();
```
Comment (vpereverzev) [inline mu4/scenes/palette/internal/paletteconfiguration.h:32]: I think it'd be better to create an interface for all configuration. They already have common things, such as the init() method or moduleName usage

File: mu4/scenes/palette/internal/paletteconfiguration.h [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+
+    double guiScale() const override;
```
Comment (vpereverzev) [inline mu4/scenes/palette/internal/paletteconfiguration.h:34]: Does this really apply to this configuration? It looks like a setting for the UI module

File: mu4/scenes/palette/palettemodule.cpp [hunk 12]
```diff
@@ -20,19 +20,30 @@
...
+#include "workspace/iworkspacedatastreamregister.h"
+#include "workspace/iworkspacemanager.h"
+#include "internal/workspacepalettestream.h"
 
 #include "libmscore/score.h"
 #include "libmscore/sym.h"
 
 using namespace mu::scene::palette;
 
+static std::shared_ptr<MU4PaletteAdapter> m_adapter = std::make_shared<MU4PaletteAdapter>();
```
Comment (vpereverzev) [inline mu4/scenes/palette/palettemodule.cpp:44]: Do we really need them to be shared?

File: mu4/scenes/palette/qml/MuseScore/Palette/Palette.qml [hunk 13]
```diff
@@ -762,12 +762,12 @@ GridView {
...
-                enabled: paletteCell.paletteDrag
-                target: mscore
-                function onElementDraggedToScoreView() { paletteCell.paletteDrag = false; }
-            }
+//--            Connections {
+//--                // force not hiding palette cell if it is being dragged to a score
+//--                enabled: paletteCell.paletteDrag
+//--                target: mscore
+//--               function onElementDraggedToScoreView() { paletteCell.paletteDrag = false; }
+//--            }
```
Comment (vpereverzev) [inline mu4/scenes/palette/qml/MuseScore/Palette/Palette.qml:770]: Are there any problems with that?

File: mu4/scenes/palette/palettemodule.cpp [hunk 14]
```diff
@@ -76,14 +105,47 @@ void PaletteModule::onInit()
...
-    gscore->style().set(Sid::MusicalTextFont, QString("Bravura Text"));
-    ScoreFont* scoreFont = ScoreFont::fontFactory("Bravura");
-    gscore->setScoreFont(scoreFont);
-    gscore->setNoteHeadWidth(scoreFont->width(SymId::noteheadBlack, gscore->spatium()) / SPATIUM20);
+    // init configuration
+    m_configuration->init();
+
+    // load workspace
+
+    auto workspaceManager = framework::ioc()->resolve<workspace::IWorkspaceManager>(moduleName());
```
Comment (vpereverzev) [inline mu4/scenes/palette/palettemodule.cpp:113]: I think this logic deserves a separate class

File: mu4/domain/notation/inotationinteraction.h [hunk 15]
```diff
@@ -64,6 +64,14 @@ class INotationInteraction
     virtual void endDrag() = 0;
     virtual async::Notification dragChanged() const = 0;
 
+    // Drop
+    //! TODO Change KeyboardModifiers to modes
+    virtual void startDrop(const QByteArray& edata) = 0;
+    virtual bool isDropAccepted(const QPointF& pos, Qt::KeyboardModifiers modifiers) = 0; //! NOTE Also may set drop target
+    virtual bool drop(const QPointF& pos, Qt::KeyboardModifiers modifiers) = 0;
+    virtual void endDrop() = 0;
+    virtual async::Notification droped() const = 0;
```
[the remark this answers is not in the document]
Comment (igorkorsukov) [reply mu4/domain/notation/inotationinteraction.h:73]: > I thought you prefer to add "Changed" word for notifiers

Yes, at first I even called `dropChanged`, but then I fixed it, it sounds ugly :)

File: mu4/domain/notation/internal/notationinteraction.h (hunk 8 above)
Comment (igorkorsukov) [reply mu4/domain/notation/internal/notationinteraction.h:107]: I can’t say, `hitElements` is trying to guess which element we hit, and `elementsAt` accurately determines. I tried not to change the logic.

File: mu4/domain/notation/notationdomainmodule.cpp (hunk 9 above)
Comment (igorkorsukov) [reply mu4/domain/notation/notationdomainmodule.cpp:50]: In this method, `MScore` is initialized, and the initialization of other subsystems depends on it. In this case, `MScore` initialization is more like export

File: mu4/scenes/palette/internal/paletteconfiguration.h (hunk 10 above)
Comment (igorkorsukov) [reply mu4/scenes/palette/internal/paletteconfiguration.h:32]: I do not want the init method to be in the interface, i.e. accessible from anywhere

File: mu4/scenes/palette/internal/paletteconfiguration.h (hunk 11 above)
Comment (igorkorsukov) [reply mu4/scenes/palette/internal/paletteconfiguration.h:34]: This method applies only to pallets, i.e. it is pallet scaling (not global)

File: mu4/scenes/palette/palettemodule.cpp (hunk 12 above)
Comment (igorkorsukov) [reply mu4/scenes/palette/palettemodule.cpp:44]: We register it in IoC, where it becomes shared, it is better that it was shared immediately

File: mu4/scenes/palette/qml/MuseScore/Palette/Palette.qml (hunk 13 above)
Comment (igorkorsukov) [reply mu4/scenes/palette/qml/MuseScore/Palette/Palette.qml]: fixed

File: mu4/scenes/palette/palettemodule.cpp (hunk 14 above)
Comment (igorkorsukov) [reply mu4/scenes/palette/palettemodule.cpp]: moved to separate class

File: framework/global/runtime.h [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+//  GNU General Public License for more details.
+//
+//  You should have received a copy of the GNU General Public License
+//  along with this program; if not, write to the Free Software
+//  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
+//=============================================================================
+#ifndef MU_FRAMEWORK_RUNTIME_H
+#define MU_FRAMEWORK_RUNTIME_H
```
Comment (RomanPudashkin) [inline framework/global/runtime.h:20]: RUNTIME

But it is calculated at compile time :)

File: framework/global/runtime.h (hunk 5 above)
Comment (RomanPudashkin) [inline framework/global/runtime.h:24]: isDebug?
